packages/flutter_tools/lib/src/android/android_studio.dart

```diff
@@ -0,0 +1,257 @@
+        configuredStudioPath = fs.path.join(configuredStudioPath, 'Contents');
+      return new AndroidStudio(configuredStudioPath,
+          configured: configuredStudio);
+    }
+
+    // Find all available Studio installations.
+    List<AndroidStudio> studios = allInstalled();
+    if (studios.isEmpty) {
+      return null;
+    }
+    studios.sort();
+    return studios.lastWhere((AndroidStudio s) => s.isValid,
+        orElse: () => null);
+  }
+
+  static List<AndroidStudio> allInstalled() =>
+      platform.isMacOS ? _allMacOS() : _allLinuxOrWindows();
+
+  static List<AndroidStudio> _allMacOS() {
+    List<FileSystemEntity> candidatePaths = <FileSystemEntity>[];
+
+    void _checkForStudio(String path) {
+      List<Directory> directories = fs
+          .directory(path)
+          .listSync()
+          .where((FileSystemEntity e) => e is Directory);
+      for (Directory directory in directories) {
+        if (directory.basename == 'Android Studio.app') {
+          candidatePaths.add(directory);
+        } else if (!directory.path.endsWith('.app')) {
+          _checkForStudio(directory.path);
+        }
+      }
+    }
+
+    _checkForStudio('/Applications');
+    _checkForStudio(fs.path.join(homeDirPath, 'Applications'));
+
+    String configuredStudioDir = config.getValue('android-studio-dir');
+    if (configuredStudioDir != null) {
+      FileSystemEntity configuredStudio = fs.file(configuredStudioDir);
+      if (configuredStudio.basename == 'Contents') {
+        configuredStudio = configuredStudio.parent;
+      }
+      if (!candidatePaths
+          .any((FileSystemEntity e) => e.path == configuredStudio.path)) {
+        candidatePaths.add(configuredStudio);
+      }
+    }
+
+    return candidatePaths
+        .map((FileSystemEntity e) => new AndroidStudio.fromMacOSBundle(e.path))
+        .where((AndroidStudio s) => s != null)
+        .toList();
+  }
+
+  static List<AndroidStudio> _allLinuxOrWindows() {
+    List<AndroidStudio> studios = <AndroidStudio>[];
+
+    bool _hasStudioAt(String path, {String newerThan}) {
+      return studios.any((AndroidStudio studio) {
+        if (studio.directory != path) return false;
+        if (newerThan != null) {
+          return studio.version.compareTo(newerThan) >= 0;
+        }
+        return true;
+      });
+    }
+
+    // Read all $HOME/AndroidStudio*/system/.home files. There may be several
+    // pointing to the same installation, so we grab only the latest one.
+    for (FileSystemEntity entity in fs.directory(homeDirPath).listSync()) {
+      if (entity is Directory && entity.basename.startsWith('.AndroidStudio')) {
+        AndroidStudio studio = new AndroidStudio.fromHomeDot(entity);
+        if (studio != null &&
+            !_hasStudioAt(studio.directory, newerThan: studio.version)) {
+          studios.removeWhere(
+              (AndroidStudio other) => other.directory == studio.directory);
```

Comment:
nit: if you put a line break after a `(`, there should be a matching one after the `)`.